### PR TITLE
fixed type and interest rate calculation

### PR DESCRIPTION
**Pull request by rudewalt**

### CODE REVIEW

File: pool/src/lib.rs [hunk 1]
```diff
@@ -767,6 +768,58 @@ impl LendingPool {
...
+    /// U - utilization, U = total_debt / total_collateral
+    /// alpha - parameter, by default 1.43 expressed as 143 with denominator 100
+    /// max_rate - maximal value of interest rate, by default 500% expressed as 50000
+    /// base_rate - base interest rate, by default 2%, expressed as 200
+    ///
+    /// For (1-U)^alpha calculation use binomial approximation with three terms
+    /// (1-U)^a = 1 - alpha * U + alpha/2 * (alpha - 1) * U^2 - alpha/6 * (alpha-1) * (alpha-2) * U^3
+    #[allow(dead_code)]
+    fn calc_interest_rate(
+        total_collateral: i128,
```
Comment (artur-abliazimov) [inline pool/src/lib.rs:783]: Should we check `total_collateral` and `total_debt` to be positive integers?

File: pool/src/lib.rs [hunk 2]
```diff
@@ -767,6 +768,58 @@ impl LendingPool {
...
+        if u >= FixedI128::ONE {
+            return None; // utilization shouldn't be greater or equal one
+        }
+
+        let alpha = FixedI128::from_rational(alpha, ALPHA_DENOMINATOR)?;
+
+        let alpha_minus_one = alpha.sub(FixedI128::ONE)?;
+        let alpha_minus_two = alpha.sub(FixedI128::from_inner(FixedI128::DENOMINATOR * 2))?;
+
+        let u_pow_2 = u.mul(u)?;
```
Comment (artur-abliazimov) [inline pool/src/lib.rs:799]: Would it be convenient to extend `FixedI128` with `pow` or `checked_pow` to replace it (and potentially other operations) with `u.pow(3)`?

File: common/src/fixedi128.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
+use fixed_point_math::FixedPoint;
+
+/// Fixed type with inner type of i128 and fixed denominator 10e9
+#[derive(Default, Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord)]
+pub struct FixedI128(i128);
+
+impl FixedI128 {
+    pub const DENOMINATOR: i128 = 1_000_000_000;
```
Comment (artur-abliazimov) [inline common/src/fixedi128.rs:8]: I see that we already have denominator in the `rate_math.rs`. Does it make sense to share the same denominator? Or maybe put it as a parameter in the `FixedI128` constructor?

File: pool/src/lib.rs (hunk 2 above)
Comment (mn13) [reply pool/src/lib.rs:799]: I think that done for efficiency - intermediate values also uses later

Comment (artur-abliazimov) [reply pool/src/lib.rs:799]: Yeah, I see it. Just wondering if it's convenient to have `pow` in `FixedI128`.

File: pool/src/lib.rs [hunk 4]
```diff
@@ -767,6 +768,58 @@ impl LendingPool {
...
+    /// where
+    /// U - utilization, U = total_debt / total_collateral
+    /// alpha - parameter, by default 1.43 expressed as 143 with denominator 100
+    /// max_rate - maximal value of interest rate, by default 500% expressed as 50000
+    /// base_rate - base interest rate, by default 2%, expressed as 200
+    ///
+    /// For (1-U)^alpha calculation use binomial approximation with three terms
+    /// (1-U)^a = 1 - alpha * U + alpha/2 * (alpha - 1) * U^2 - alpha/6 * (alpha-1) * (alpha-2) * U^3
+    #[allow(dead_code)]
+    fn calc_interest_rate(
```
Comment (artur-abliazimov) [inline pool/src/lib.rs:782]: As we discussed on the meeting, let's also add some tests to verify it works as expected under usual and unusual circumstances.

File: common/src/fixedi128.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    /// Div inner value of fixed
+    pub fn div_inner<T: Into<i128>>(self, value: T) -> Option<FixedI128> {
+        self.0.checked_div(value.into()).map(FixedI128)
+    }
+}
+
+#[cfg(test)]
+mod test {
```
Comment (artur-abliazimov) [inline common/src/fixedi128.rs:96]: Could you move this test to a separate file?

File: common/src/fixedi128.rs (hunk 3 above)
Comment (rudewalt) [reply common/src/fixedi128.rs:10]: rate_math.rs was previous way to work with fixed and should be replaced by FixedI128 and deleted

Comment (artur-abliazimov) [reply common/src/fixedi128.rs:10]: Ok, sounds good to me.

Comment (rudewalt) [reply common/src/fixedi128.rs:10]: In the next PR

**Pull request merged**